### PR TITLE
A request to enable reading zlog configurations from run-time memory

### PR DESCRIPTION
**Pull request by OverDriveGain**
**Requirement**
I am using Zlog as the primary logging library of one project I am working on. The library misses one feature which is expected to be included. In Zlog we cannot read the configurations of the log from run-time memory variables, we can only read them from configuration file. e.g. like follows:

`char * config = "[formats]\n default_format = \"%m\" \n" ";`
`zlog_init(config);`
or 
`zlog_reload(config);`

**Changes in code**
I have extended the code of your library and included this new feature, zlog can now be configured and initiated as mentioned above. Since the library could already read from file, including this new feature was straight forward. The steps were taken was adding only one function to directly read the lines from the buffer and thus bypassing the function of reading from the file descriptor. 

**Behaviour changes**
The user now has to define a lock file explicitly since there is no log file to be used as a lock file.

**Testing**
I have tested the new code completely and made sure no errors would occur. Actually the code has been used with the development of one project since more than six months, this means all issues that might ever happen have been tracked and resolved. 

Thank you so much for writing this library and thank you for accepting my previous merge request.

### CODE REVIEW

File: src/conf.c [hunk 1]
```diff
@@ -106,15 +114,20 @@ zlog_conf_t *zlog_conf_new(const char *confpath)
...
+	if (config && config[0] != '\0' && config[0] != '[') {
+		nwrite = snprintf(a_conf->file, sizeof(a_conf->file), "%s", config);
+		cfg_source = FILE_CFG;
 	} else if (getenv("ZLOG_CONF_PATH") != NULL) {
 		nwrite = snprintf(a_conf->file, sizeof(a_conf->file), "%s", getenv("ZLOG_CONF_PATH"));
-		has_conf_file = 1;
-	} else {
+		cfg_source = FILE_CFG;
+	} else if (config[0]=='[') {
+        memset(a_conf->file, 0x00, sizeof(a_conf->file));
```
Comment (deemic) [inline src/conf.c]: Here should add indent

File: src/conf.c [hunk 2]
```diff
@@ -155,11 +168,16 @@ zlog_conf_t *zlog_conf_new(const char *confpath)
...
 	}
 
-	if (has_conf_file) {
+	if (cfg_source == FILE_CFG) {
 		if (zlog_conf_build_with_file(a_conf)) {
 			zc_error("zlog_conf_build_with_file fail");
 			goto err;
 		}
+	} else if (cfg_source == IN_MEMORY_CFG) {
+				if(zlog_conf_build_with_in_memory(a_conf)){
```
Comment (deemic) [inline src/conf.c]: wrong indent

File: src/conf.c [hunk 3]
```diff
@@ -327,7 +345,33 @@ static int zlog_conf_build_with_file(zlog_conf_t * a_conf)
...
+	int section = 0;
+	pline = line;
+	memset(&line, 0x00, sizeof(line));
+	pline = strtok((char *)a_conf->cfg_ptr, "\n");
 
+	while (pline != NULL) {
+		rc = zlog_conf_parse_line(a_conf, pline, &section);
+		if (rc < 0) {
+			zc_error("parse in-memory configurations[%s] line [%s] fail", a_conf->cfg_ptr, pline);
+			goto exit;
```
Comment (deemic) [inline src/conf.c]: No need to use `goto`, a `break` will be just ok

**Comment by deemic** [comment]
I think a test case is necessary.

**Comment by OverDriveGain** [comment]
OK, I have made another test.

**Comment by OverDriveGain** [comment]
Hello, can you please consider merging my request?

**Comment by OverDriveGain** [comment]
Since the pull request was approved, I suppose my commits can be merged now?

**Pull request merged**